<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Biology 304: Biological Data Analysis</title>
  <meta name="description" content="Hands-on materials for Bio 304 course at Duke University.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Biology 304: Biological Data Analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 304 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Biology 304: Biological Data Analysis" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 304 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2018-09-20">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction-to-the-dplyr-package.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html"><i class="fa fa-check"></i><b>2</b> Data story: Women and children first on the Titanic?</a><ul>
<li class="chapter" data-level="2.1" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#background"><i class="fa fa-check"></i><b>2.1</b> Background</a></li>
<li class="chapter" data-level="2.2" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#dataset"><i class="fa fa-check"></i><b>2.2</b> Dataset</a></li>
<li class="chapter" data-level="2.3" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#libraries"><i class="fa fa-check"></i><b>2.3</b> Libraries</a></li>
<li class="chapter" data-level="2.4" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#read-data"><i class="fa fa-check"></i><b>2.4</b> Read data</a></li>
<li class="chapter" data-level="2.5" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#whats-in-the-data"><i class="fa fa-check"></i><b>2.5</b> What’s in the data?</a><ul>
<li class="chapter" data-level="2.5.1" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#simple-data-wrangling"><i class="fa fa-check"></i><b>2.5.1</b> Simple data wrangling</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#categorizing-passengers"><i class="fa fa-check"></i><b>2.6</b> Categorizing passengers</a><ul>
<li class="chapter" data-level="2.6.1" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#how-many-people-survived"><i class="fa fa-check"></i><b>2.6.1</b> How many people survived?</a></li>
<li class="chapter" data-level="2.6.2" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#women-first"><i class="fa fa-check"></i><b>2.6.2</b> Women first?</a></li>
<li class="chapter" data-level="2.6.3" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#a-bar-plot-using-proportions-rather-than-counts"><i class="fa fa-check"></i><b>2.6.3</b> A bar plot using proportions rather than counts</a></li>
<li class="chapter" data-level="2.6.4" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#mosaic-plots"><i class="fa fa-check"></i><b>2.6.4</b> Mosaic plots</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#passenger-ages"><i class="fa fa-check"></i><b>2.7</b> Passenger ages</a></li>
<li class="chapter" data-level="2.8" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#how-does-age-relate-to-survival"><i class="fa fa-check"></i><b>2.8</b> How does age relate to survival?</a><ul>
<li class="chapter" data-level="2.8.1" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#strip-charts"><i class="fa fa-check"></i><b>2.8.1</b> Strip charts</a></li>
<li class="chapter" data-level="2.8.2" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#box-plots"><i class="fa fa-check"></i><b>2.8.2</b> Box plots</a></li>
<li class="chapter" data-level="2.8.3" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#fitting-a-model-relating-survival-to-age-and-sex"><i class="fa fa-check"></i><b>2.8.3</b> Fitting a model relating survival to age and sex</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#how-does-class-affect-survival"><i class="fa fa-check"></i><b>2.9</b> How does class affect survival?</a><ul>
<li class="chapter" data-level="2.9.1" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#combining-sex-and-class"><i class="fa fa-check"></i><b>2.9.1</b> Combining Sex and Class</a></li>
<li class="chapter" data-level="2.9.2" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#extending-the-logistic-regression-model-to-consider-class-age-sex-and-survival"><i class="fa fa-check"></i><b>2.9.2</b> Extending the logistic regression model to consider class, age, sex, and survival</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="data-story-women-and-children-first-on-the-titanic.html"><a href="data-story-women-and-children-first-on-the-titanic.html#conclusion"><i class="fa fa-check"></i><b>2.10</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>3</b> Getting Started with R</a><ul>
<li class="chapter" data-level="3.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>3.1</b> What is R?</a></li>
<li class="chapter" data-level="3.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>3.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="3.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>3.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="3.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>3.4</b> Comments</a></li>
<li class="chapter" data-level="3.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>3.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="3.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>3.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>3.6</b> The R Help System</a><ul>
<li class="chapter" data-level="3.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>3.6.1</b> Getting help from the console</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment-in-r"><i class="fa fa-check"></i><b>3.7</b> Variable assignment in R</a><ul>
<li class="chapter" data-level="3.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>3.7.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>3.8</b> Data types</a></li>
<li class="chapter" data-level="3.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>3.9</b> Logical values</a><ul>
<li class="chapter" data-level="3.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-operators"><i class="fa fa-check"></i><b>3.9.1</b> Logical operators</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>3.10</b> Character strings</a><ul>
<li class="chapter" data-level="3.10.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#joining-strings"><i class="fa fa-check"></i><b>3.10.1</b> Joining strings</a></li>
<li class="chapter" data-level="3.10.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#splitting-strings"><i class="fa fa-check"></i><b>3.10.2</b> Splitting strings</a></li>
<li class="chapter" data-level="3.10.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#substrings"><i class="fa fa-check"></i><b>3.10.3</b> Substrings</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>3.11</b> Packages</a><ul>
<li class="chapter" data-level="3.11.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>3.11.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="3.11.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>3.11.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="3.11.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>3.11.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="3.11.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>3.11.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="3.12" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#reading-data-from-a-file"><i class="fa fa-check"></i><b>3.12</b> Reading data from a file</a></li>
<li class="chapter" data-level="3.13" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#exploring-the-possums-data-set"><i class="fa fa-check"></i><b>3.13</b> Exploring the “possums” data set</a></li>
<li class="chapter" data-level="3.14" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-tables"><i class="fa fa-check"></i><b>3.14</b> Simple tables</a></li>
<li class="chapter" data-level="3.15" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-figures"><i class="fa fa-check"></i><b>3.15</b> Simple figures</a></li>
<li class="chapter" data-level="3.16" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#bar-plots"><i class="fa fa-check"></i><b>3.16</b> Bar plots</a></li>
<li class="chapter" data-level="3.17" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#histograms"><i class="fa fa-check"></i><b>3.17</b> Histograms</a></li>
<li class="chapter" data-level="3.18" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#scatter-plots"><i class="fa fa-check"></i><b>3.18</b> Scatter plots</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>4</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>4.1</b> R Notebooks</a></li>
<li class="chapter" data-level="4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>4.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="4.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>4.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="4.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>4.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>4.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="4.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>4.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>4.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="4.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>4.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="4.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>4.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="4.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>4.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html"><i class="fa fa-check"></i><b>5</b> More R Basics: Data structures</a><ul>
<li class="chapter" data-level="5.1" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#vectors"><i class="fa fa-check"></i><b>5.1</b> Vectors</a><ul>
<li class="chapter" data-level="5.1.1" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>5.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="5.1.2" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>5.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="5.1.3" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>5.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="5.1.4" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>5.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="5.1.5" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>5.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="5.1.6" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>5.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="5.1.7" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>5.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="5.1.8" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>5.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="5.1.9" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>5.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#lists"><i class="fa fa-check"></i><b>5.2</b> Lists</a><ul>
<li class="chapter" data-level="5.2.1" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>5.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="5.2.2" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>5.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="5.2.3" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>5.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="5.2.4" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#the-operator"><i class="fa fa-check"></i><b>5.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="5.2.5" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>5.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="5.2.6" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#combining-lists"><i class="fa fa-check"></i><b>5.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="5.2.7" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>5.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#data-frames"><i class="fa fa-check"></i><b>5.3</b> Data frames</a><ul>
<li class="chapter" data-level="5.3.1" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>5.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="5.3.2" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>5.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="5.3.3" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>5.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="5.3.4" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>5.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="5.3.5" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>5.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="5.3.6" data-path="more-r-basics-data-structures.html"><a href="more-r-basics-data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>5.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>6</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>6.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>6.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>6.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>6.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>6.5</b> Strip plots</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>6.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>6.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>6.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms-1"><i class="fa fa-check"></i><b>6.6</b> Histograms</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>6.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>6.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="6.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>6.8</b> Density plots</a></li>
<li class="chapter" data-level="6.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>6.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="6.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>6.10</b> Boxplots</a></li>
<li class="chapter" data-level="6.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>6.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="6.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>6.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="6.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>6.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="6.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>6.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="6.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="6.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>6.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="6.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>6.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="6.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>6.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="6.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>6.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="6.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="6.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>6.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="6.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots-1"><i class="fa fa-check"></i><b>6.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="6.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>6.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="6.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>6.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="6.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>6.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="6.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>6.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="6.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>6.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="6.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>6.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html"><i class="fa fa-check"></i><b>7</b> Introduction to the <code>dplyr</code> package</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>7.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#example-data-nc-births"><i class="fa fa-check"></i><b>7.2.1</b> Example data: NC Births</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-excel-files"><i class="fa fa-check"></i><b>7.2.2</b> Reading Excel files</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>7.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="7.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>7.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="7.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#dplyrs-verbs"><i class="fa fa-check"></i><b>7.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#select"><i class="fa fa-check"></i><b>7.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#filter"><i class="fa fa-check"></i><b>7.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#mutate"><i class="fa fa-check"></i><b>7.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="7.5.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#arrange"><i class="fa fa-check"></i><b>7.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="7.5.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#summarize"><i class="fa fa-check"></i><b>7.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#group_by"><i class="fa fa-check"></i><b>7.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="7.5.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>7.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="7.5.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>7.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>7.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#pipes"><i class="fa fa-check"></i><b>7.6</b> Pipes</a><ul>
<li class="chapter" data-level="7.6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>7.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>7.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="7.6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#an-example-without-pipes"><i class="fa fa-check"></i><b>7.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="7.6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>7.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="7.6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>7.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="7.6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>7.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="7.6.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>7.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="7.6.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>7.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html"><i class="fa fa-check"></i><b>8</b> Data wrangling, Part I</a><ul>
<li class="chapter" data-level="8.1" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#libraries-2"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#data"><i class="fa fa-check"></i><b>8.2</b> Data</a></li>
<li class="chapter" data-level="8.3" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>8.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="8.4" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>8.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="8.4.1" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>8.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="8.4.2" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>8.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#merging-data-frames"><i class="fa fa-check"></i><b>8.5</b> Merging data frames</a></li>
<li class="chapter" data-level="8.6" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>8.6</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="8.6.1" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>8.6.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="8.6.2" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>8.6.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="8.6.3" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#subsetting-rows"><i class="fa fa-check"></i><b>8.6.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="8.6.4" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#splitting-columns"><i class="fa fa-check"></i><b>8.6.4</b> Splitting columns</a></li>
<li class="chapter" data-level="8.6.5" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#combining-data-frames"><i class="fa fa-check"></i><b>8.6.5</b> Combining data frames</a></li>
<li class="chapter" data-level="8.6.6" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>8.6.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#using-your-tidy-data"><i class="fa fa-check"></i><b>8.7</b> Using your tidy data</a><ul>
<li class="chapter" data-level="8.7.1" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>8.7.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="8.7.2" data-path="data-wrangling-part-i.html"><a href="data-wrangling-part-i.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>8.7.2</b> Finding the most variable genes</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 304: Biological Data Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-wrangling-part-i" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Data wrangling, Part I</h1>
<p>In the real world you’ll often create a data set (or be given one) in a format that is less than ideal for analysis. This can happen for a number of reasons. For example, the data may have been recorded in a manner convenient for collection and visual inspection, but which does not work well for analysis and plotting. Or the data may be an amalgamation of multiple experiments, in which each of the experimenters used slightly different naming conventions. Or the data may have been produced by an instrument that produces output with a fixed format. Sometimes important experimental information is included in the column headers of a spreadsheet.</p>
<p>Whatever the case, we often find ourselves in the situation where we need to “wrangle” our data into a “tidy” format before we can proceed with visualization and analysis. The “R for Data Science” text discusses some desirable rules for “tidy” data in order to facilitate downstream analyses. These are:</p>
<ol style="list-style-type: decimal">
<li>Each variable must have its own column.</li>
<li>Each observation must have its own row.</li>
<li>Each value must have its own cell.</li>
</ol>
<p>In this lecture we’re going to walk through an extended example of wrangling some data into a “tidy” format.</p>
<div id="libraries-2" class="section level2">
<h2><span class="header-section-number">8.1</span> Libraries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(magrittr)
<span class="kw">library</span>(stringr)
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(cowplot)</code></pre></div>
</div>
<div id="data" class="section level2">
<h2><span class="header-section-number">8.2</span> Data</h2>
<p>To illustrate a data wrangling pipeline, we’re going to use a gene expression microarray data set, based on the following paper:</p>
<ul>
<li>Spellman PT, et al. 1998. Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. Mol Biol Cell 9(12): 3273-97.</li>
</ul>
<p>In this paper, Spellman and colleagues tried to identify all the genes in the yeast genome (&gt;6000 genes) that exhibited oscillatory behaviors suggestive of cell cycle regulation. To do so, they combined gene expression measurements from six different types of cell cycle synchronization experiments.</p>
<p><em>Download the Spellman data to your filesystem</em> from <a href="https://github.com/bio304-class/bio304-course-notes/raw/master/datasets/spellman-combined.txt">this link</a> (right-click the “Download” button and save to your Downloads folder or similar).</p>
<p>I suggest that once you download the data, you open it in a spreadsheet program (e.g. Excel) or use the RStudio Data Viewer to get a sense of what the data looks like.</p>
<p>Let’s load it into R, using the <code>read_tsv()</code> function, using the appropriate file path.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the filepath may differ on your computer</span>
spellman &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;~/Downloads/spellman-combined.txt&quot;</span>)
<span class="co">#&gt; Parsed with column specification:</span>
<span class="co">#&gt; cols(</span>
<span class="co">#&gt;   .default = col_double(),</span>
<span class="co">#&gt;   X1 = col_character(),</span>
<span class="co">#&gt;   clb = col_character(),</span>
<span class="co">#&gt;   alpha = col_character(),</span>
<span class="co">#&gt;   cdc15 = col_character(),</span>
<span class="co">#&gt;   cdc28 = col_character(),</span>
<span class="co">#&gt;   elu = col_character()</span>
<span class="co">#&gt; )</span>
<span class="co">#&gt; See spec(...) for full column specifications.</span></code></pre></div>
<p>The initial dimenions of the data frame are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman)
<span class="co">#&gt; [1] 6178   83</span></code></pre></div>
<p>The six types of cell cycle synchronization experiments included in this data set are:</p>
<ol style="list-style-type: decimal">
<li>synchronization by alpha-factor = “alpha”</li>
<li>synchronization by cdc15 temperature sensitive mutants = “cdc15”</li>
<li>synchronization by cdc28 temperature sensitive mutants = “cdc28”</li>
<li>synchronization by elutration = “elu”</li>
<li>synchronization by cln3 mutatant strains = “cln3”</li>
<li>synchronization by clb2 mutant strains = “clb2”</li>
</ol>
</div>
<div id="renaming-data-frame-columms" class="section level2">
<h2><span class="header-section-number">8.3</span> Renaming data frame columms</h2>
<p>Notice that when we imported the data we got a warning message: <code>Missing column names filled in: 'X1' [1]</code>. In a data frame, every column must have a name. The first column of our data set did not have a name in the header, so <code>read_tsv</code> automatically gave it the name <code>X1</code>.</p>
<p>Our first task is to give the first column a more meaningful name. This column gives “systematic gene names” – a standardized naming scheme for genes in the yeast genome. We’ll use <code>dplyr::rename</code> to rename <code>X1</code> to <code>gene</code>. Note that <code>rename</code> can take multiple arguments if you need to rename multiple columns simultaneously.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean &lt;-<span class="st"> </span>
<span class="st">  </span>spellman <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rename</span>(<span class="dt">gene =</span> X1)</code></pre></div>
<p>Note the use of the compound assingment operator – <code>%&lt;&gt;%</code> – from the <code>magrittr</code> package, which we introduced in our last class session.</p>
</div>
<div id="dropping-unneeded-columns" class="section level2">
<h2><span class="header-section-number">8.4</span> Dropping unneeded columns</h2>
<p>Take a look at the Spellman data again in your spreadsheet program (or the RStudio data viewer). You’ll notice there are some blank columns. For example there is a column with the header “alpha” that has no entries. These are simply visual organizing elements that the creator of the spreadsheet added to separate the different experiments that are included in the data set.</p>
<p>We can use <code>dplyr::select()</code> to drop columns by prepending column names with the negative sign:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># drop the alpha column keeping all others</span>
spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>alpha)  </code></pre></div>
<p>Note that usually <code>select()</code> keeps only the variables you specify. However if the first expression is negative, <code>select</code> will instead automatically keep all variables, dropping only those you specify.</p>
<div id="finding-all-empty-columns" class="section level3">
<h3><span class="header-section-number">8.4.1</span> Finding all empty columns</h3>
<p>In the example above, we looked at the data and saw that the “alpha” column was empty, and thus dropped it. This worked because there are only a modest number of columns in the data frame in it’s initial form. However, if our data frame contained thousands of columns, this “look and see” procedure would not be efficient. Can we come up with a general solution for removing empty columns from a data frame?</p>
<p>When you load a data frame from a spreadsheet, empty cells are given the value <code>NA</code>. In previous class sessions we were introduced to the function <code>is.na()</code> which tests each value in a vector or data frame for whether it’s NA or not. We can count NA values in a vector by summing the output of <code>is.na()</code>. Conversely we can count the number of “not NA” items by using the negation operator (<code>!</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># count number of NA values in the alpha0 column</span>
<span class="kw">sum</span>(<span class="kw">is.na</span>(spellman<span class="op">$</span>alpha0))
<span class="co">#&gt; [1] 165</span>

<span class="co"># count number of values that are NOT NA in alpha0</span>
<span class="kw">sum</span>(<span class="op">!</span><span class="kw">is.na</span>(spellman<span class="op">$</span>alpha0))
<span class="co">#&gt; [1] 6013</span></code></pre></div>
<p>This seems like it should get us close to a solution but <code>sum(is.na(..))</code> when applied to a data frame counts NAs across the entire data frame, not column-by-column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># doesn&#39;t do what we hoped!</span>
<span class="kw">sum</span>(<span class="kw">is.na</span>(spellman))
<span class="co">#&gt; [1] 59017</span></code></pre></div>
<p>If we want sums of NAs by column, we instead use the <code>colSums()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get number of NAs by column</span>
<span class="kw">colSums</span>(<span class="kw">is.na</span>(spellman))
<span class="co">#&gt;        X1    cln3-1    cln3-2       clb    clb2-2    clb2-1     alpha </span>
<span class="co">#&gt;         0       193       365      6178       454       142      6178 </span>
<span class="co">#&gt;    alpha0    alpha7   alpha14   alpha21   alpha28   alpha35   alpha42 </span>
<span class="co">#&gt;       165       525       191       312       267       207       123 </span>
<span class="co">#&gt;   alpha49   alpha56   alpha63   alpha70   alpha77   alpha84   alpha91 </span>
<span class="co">#&gt;       257       147       186       185       178       155       329 </span>
<span class="co">#&gt;   alpha98  alpha105  alpha112  alpha119     cdc15  cdc15_10  cdc15_30 </span>
<span class="co">#&gt;       209       174       222       251      6178       677       477 </span>
<span class="co">#&gt;  cdc15_50  cdc15_70  cdc15_80  cdc15_90 cdc15_100 cdc15_110 cdc15_120 </span>
<span class="co">#&gt;       501       608       573       562       606       570       611 </span>
<span class="co">#&gt; cdc15_130 cdc15_140 cdc15_150 cdc15_160 cdc15_170 cdc15_180 cdc15_190 </span>
<span class="co">#&gt;       495       574       811       583       571       803       613 </span>
<span class="co">#&gt; cdc15_200 cdc15_210 cdc15_220 cdc15_230 cdc15_240 cdc15_250 cdc15_270 </span>
<span class="co">#&gt;      1014       573       741       596       847       379       537 </span>
<span class="co">#&gt; cdc15_290     cdc28   cdc28_0  cdc28_10  cdc28_20  cdc28_30  cdc28_40 </span>
<span class="co">#&gt;       426      6178       122        72        67        55        66 </span>
<span class="co">#&gt;  cdc28_50  cdc28_60  cdc28_70  cdc28_80  cdc28_90 cdc28_100 cdc28_110 </span>
<span class="co">#&gt;        56        82        84        75       237       165       319 </span>
<span class="co">#&gt; cdc28_120 cdc28_130 cdc28_140 cdc28_150 cdc28_160       elu      elu0 </span>
<span class="co">#&gt;       312      1439      2159       521       543      6178       122 </span>
<span class="co">#&gt;     elu30     elu60     elu90    elu120    elu150    elu180    elu210 </span>
<span class="co">#&gt;       153       175       132       103       119       111       118 </span>
<span class="co">#&gt;    elu240    elu270    elu300    elu330    elu360    elu390 </span>
<span class="co">#&gt;       131       110       112       112       156       114</span></code></pre></div>
<p>Columns with <em>all</em> missing values can be more conveniently found by asking for those columns where the number of “not missing” values is zero:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get names of all columns for which all rows are NA</span>
<span class="co"># useing standard indexing</span>
<span class="kw">names</span>(spellman)[<span class="kw">colSums</span>(<span class="op">!</span><span class="kw">is.na</span>(spellman)) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>]
<span class="co">#&gt; [1] &quot;clb&quot;   &quot;alpha&quot; &quot;cdc15&quot; &quot;cdc28&quot; &quot;elu&quot;</span></code></pre></div>
<p>We can combine the <code>colSums(!is.na())</code> idiom with the <code>dplyr::select_if</code> function to quickly remove all empty columns as so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="co"># keep ONLY the non-empty columns</span>
<span class="st">  </span><span class="kw">select_if</span>(<span class="kw">colSums</span>(<span class="op">!</span><span class="kw">is.na</span>(.)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</code></pre></div>
</div>
<div id="dropping-columns-by-matching-names" class="section level3">
<h3><span class="header-section-number">8.4.2</span> Dropping columns by matching names</h3>
<p>Only two time points from the cln3 and clb2 experiments were reported in the original publication. Since complete time series are unavailable for these two experimental conditions we will drop them from further consideration.</p>
<p><code>select()</code> can be called be called with a number of “helper function” (<code>?select_helpers</code>). Here we’ll illustrate the <code>matches()</code> helper function which matches column names to a “regular expression”. Regular expressions (also referred to as “regex” or “regexp”) are a way of specifying patterns in strings. For the purposes of this document we’ll illustrate regexs by example; for a more detailed explanation of regular expressions see the the regex help(<code>?regex</code>) and the <a href="http://r4ds.had.co.nz/strings.html">Chapter on Strings in “R for Data Analysis”</a>:</p>
<p>Let’s see how to drop all the “cln3” and “clb2” columns from the data frame using <code>matches()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;cln3&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;clb2&quot;</span>))</code></pre></div>
<p>If we wanted we could have collapsed our two match statements into one as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;cln3|clb2&quot;</span>))</code></pre></div>
<p>In this second example, the character “|” is specifing an OR match within the regular expression, so this regular expression matches column names that contain “cln3” OR “clb2”.</p>
</div>
</div>
<div id="merging-data-frames" class="section level2">
<h2><span class="header-section-number">8.5</span> Merging data frames</h2>
<p>Often you’ll find yourself in the situation where you want to combine information from multiple data sources. The usual requirement is that the data sources have one or more shared columns, that allow you to relate the entities or observations (rows) between the data sets. <code>dplyr</code> provides a variety of <code>join</code> functions to handle different data merging operators.</p>
<p>To illustrating merging or joining data sources, we’ll add information about each genes “common name” and a description of the gene functions to our Spellman data set. I’ve prepared a file with this info based on info I downloaded from the <a href="https://www.yeastgenome.org">Saccharomyces Genome Database</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.info &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;https://github.com/bio304-class/bio304-course-notes/raw/master/datasets/yeast-ORF-info.csv&quot;</span>)
<span class="co">#&gt; Parsed with column specification:</span>
<span class="co">#&gt; cols(</span>
<span class="co">#&gt;   ftr.name = col_character(),</span>
<span class="co">#&gt;   std.name = col_character(),</span>
<span class="co">#&gt;   description = col_character()</span>
<span class="co">#&gt; )</span></code></pre></div>
<p>Having loaded the data, let’s get a quick overview of it’s structure:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(gene.info)
<span class="co">#&gt; [1] &quot;ftr.name&quot;    &quot;std.name&quot;    &quot;description&quot;</span>
<span class="kw">dim</span>(gene.info)
<span class="co">#&gt; [1] 6610    3</span></code></pre></div>
<p>In <code>gene.info</code>, the <code>ftr.name</code> column corresponds to the <code>gene</code> column in our Spellman data set. The <code>std.name</code> column gives the “common” gene name (not every gene has a common name so there are lots of NAs). The <code>description</code> column gives a brief textual description of what the gene product does.</p>
<p>To combine <code>spellmean.clean</code> with <code>gene.info</code> we use the <code>left_join</code> function defined in dplyr. As noted in the description of the function, <code>left_join(x, y)</code> returns “all rows from x, and all columns from x and y. Rows in x with no match in y will have NA values in the new columns.” In addition, we have to specify the column to join by using the <code>by</code> argument to <code>left_join</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.merged &lt;-
<span class="st">  </span><span class="kw">left_join</span>(spellman.clean, gene.info, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;gene&quot;</span> =<span class="st"> &quot;ftr.name&quot;</span>))</code></pre></div>
<p>By default, the joined columns are merged at the end of the data frame, so we’ll reorder variables to bring the <code>std.name</code> and <code>description</code> to the second and thirds columns, preserving the order of all the other colums.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.merged <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(gene, std.name, description, <span class="kw">everything</span>())

spellman.merged
<span class="co">#&gt; # A tibble: 6,178 x 76</span>
<span class="co">#&gt;    gene  std.name description alpha0 alpha7 alpha14 alpha21 alpha28 alpha35</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 YAL0… TFC3     Subunit of…  -0.15  -0.15   -0.21    0.17   -0.42  -0.44 </span>
<span class="co">#&gt;  2 YAL0… VPS8     Membrane-b…  -0.11   0.1     0.01    0.06    0.04  -0.26 </span>
<span class="co">#&gt;  3 YAL0… EFB1     Translatio…  -0.14  -0.71    0.1    -0.32   -0.4   -0.580</span>
<span class="co">#&gt;  4 YAL0… &lt;NA&gt;     Dubious op…  -0.02  -0.48   -0.11    0.12   -0.03   0.19 </span>
<span class="co">#&gt;  5 YAL0… SSA1     ATPase inv…  -0.05  -0.53   -0.47   -0.06    0.11  -0.07 </span>
<span class="co">#&gt;  6 YAL0… ERP2     Member of …  -0.6   -0.45   -0.13    0.35   -0.01   0.49 </span>
<span class="co">#&gt;  7 YAL0… FUN14    Integral m…  -0.28  -0.22   -0.06    0.22    0.25   0.13 </span>
<span class="co">#&gt;  8 YAL0… SPO7     Putative r…  -0.03  -0.27    0.17   -0.12   -0.27   0.06 </span>
<span class="co">#&gt;  9 YAL0… MDM10    Subunit of…  -0.05   0.13    0.13   -0.21   -0.45  -0.21 </span>
<span class="co">#&gt; 10 YAL0… SWC3     Protein of…  -0.31  -0.43   -0.3    -0.23   -0.13  -0.07 </span>
<span class="co">#&gt; # ... with 6,168 more rows, and 67 more variables: alpha42 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   alpha49 &lt;dbl&gt;, alpha56 &lt;dbl&gt;, alpha63 &lt;dbl&gt;, alpha70 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   alpha77 &lt;dbl&gt;, alpha84 &lt;dbl&gt;, alpha91 &lt;dbl&gt;, alpha98 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   alpha105 &lt;dbl&gt;, alpha112 &lt;dbl&gt;, alpha119 &lt;dbl&gt;, cdc15_10 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc15_30 &lt;dbl&gt;, cdc15_50 &lt;dbl&gt;, cdc15_70 &lt;dbl&gt;, cdc15_80 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc15_90 &lt;dbl&gt;, cdc15_100 &lt;dbl&gt;, cdc15_110 &lt;dbl&gt;, cdc15_120 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc15_130 &lt;dbl&gt;, cdc15_140 &lt;dbl&gt;, cdc15_150 &lt;dbl&gt;, cdc15_160 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc15_170 &lt;dbl&gt;, cdc15_180 &lt;dbl&gt;, cdc15_190 &lt;dbl&gt;, cdc15_200 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc15_210 &lt;dbl&gt;, cdc15_220 &lt;dbl&gt;, cdc15_230 &lt;dbl&gt;, cdc15_240 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc15_250 &lt;dbl&gt;, cdc15_270 &lt;dbl&gt;, cdc15_290 &lt;dbl&gt;, cdc28_0 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc28_10 &lt;dbl&gt;, cdc28_20 &lt;dbl&gt;, cdc28_30 &lt;dbl&gt;, cdc28_40 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc28_50 &lt;dbl&gt;, cdc28_60 &lt;dbl&gt;, cdc28_70 &lt;dbl&gt;, cdc28_80 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc28_90 &lt;dbl&gt;, cdc28_100 &lt;dbl&gt;, cdc28_110 &lt;dbl&gt;, cdc28_120 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cdc28_130 &lt;dbl&gt;, cdc28_140 &lt;dbl&gt;, cdc28_150 &lt;dbl&gt;, cdc28_160 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   elu0 &lt;dbl&gt;, elu30 &lt;dbl&gt;, elu60 &lt;dbl&gt;, elu90 &lt;dbl&gt;, elu120 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   elu150 &lt;dbl&gt;, elu180 &lt;dbl&gt;, elu210 &lt;dbl&gt;, elu240 &lt;dbl&gt;, elu270 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   elu300 &lt;dbl&gt;, elu330 &lt;dbl&gt;, elu360 &lt;dbl&gt;, elu390 &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="reshaping-data-with-tidyr" class="section level2">
<h2><span class="header-section-number">8.6</span> Reshaping data with tidyr</h2>
<p>The <code>tidyr</code> package provides functions for reshaping or tidying data frames. <code>tidyr</code> is yet another component of the tidyverse, and thus was loaded by the <code>library(tidyverse)</code>.</p>
<p>We’re going to look at two functions <code>tidyr::gather()</code> and <code>tidyr::extract()</code>, and how they can be combined with now familiar <code>dplyr</code> functions we’ve seen previously. The reading assignment for today’s class session covers a variety of other functions defined in <code>tidyr</code>.</p>
<p>The Spellman data, as I provided it to you, is in what we would call “wide” format. Each column (besides the <code>gene</code> column) corresponds to an experimental condition <em>and</em> time point. For example, “alpha0” is the alpha-factor experiment at time point 0 mins; “alpha7” is the alpha-factor experiment at time point 7 mins, etc. The cells within each column correspond to the expression of a corresponding gene (given by the first column which we renamed <code>gene</code>) in that particular experiment at that particular time point.</p>
<p>In every column (except “gene”), the cells represents the same abstract property of interest – the expression of a gene of interest in a particular experiment/time point. Our first task will be to rearrange our “wide” data frame that consists of many different columns representing gene expression into a “long” data frame with just a single column representing expression. We’ll also create a new column to keep track of which experiment and time point the measurement came from.</p>
<div id="wide-to-long-conversions-using-tidyrgather" class="section level3">
<h3><span class="header-section-number">8.6.1</span> Wide to long conversions using <code>tidyr::gather</code></h3>
<p><code>tidyr::gather()</code> takes multiple columns, and collapses them together into a smaller number of new columns. When using <code>gather()</code> you give the names of the <em>new</em> columns to create, as well as the names of any existing columns <code>gather()</code> should <em>not</em> collect together.</p>
<p>Here we want to collapse all 73 or the expression columns – “alpha0” to “elu390” – into two columns: 1) a column to represent the expt/time point of the measurement, and 2) a column to represent the corresponding expression value. The column we don’t want to touch are the <code>gene</code>, <code>std.name</code>, and <code>description</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert &quot;wide&quot; data to &quot;long&quot;</span>
spellman.long &lt;-
<span class="st">  </span>spellman.merged <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">gather</span>(expt.and.time, expression, <span class="op">-</span>gene, <span class="op">-</span>std.name, <span class="op">-</span>description)</code></pre></div>
<p>Take a moment to look at the data in the “long format”:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(spellman.long)
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   gene   std.name description                     expt.and.time expression</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                           &lt;chr&gt;              &lt;dbl&gt;</span>
<span class="co">#&gt; 1 YAL00… TFC3     Subunit of RNA polymerase III … alpha0             -0.15</span>
<span class="co">#&gt; 2 YAL00… VPS8     Membrane-binding component of … alpha0             -0.11</span>
<span class="co">#&gt; 3 YAL00… EFB1     Translation elongation factor … alpha0             -0.14</span>
<span class="co">#&gt; 4 YAL00… &lt;NA&gt;     Dubious open reading frame; un… alpha0             -0.02</span>
<span class="co">#&gt; 5 YAL00… SSA1     ATPase involved in protein fol… alpha0             -0.05</span>
<span class="co">#&gt; 6 YAL00… ERP2     Member of the p24 family invol… alpha0             -0.6</span></code></pre></div>
<p>And compare the dimensions of the wide data to the new data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman.merged)  <span class="co"># for comparison</span>
<span class="co">#&gt; [1] 6178   76</span>
<span class="kw">dim</span>(spellman.long)
<span class="co">#&gt; [1] 450994      5</span></code></pre></div>
<p>As you see, we’ve gone from a data frame with 6178 rows and 76 columns (wide format), to a new data frame with 450994 rows and 5 columns (long format).</p>
</div>
<div id="extracting-information-from-combined-variables-using-tidyrextract" class="section level3">
<h3><span class="header-section-number">8.6.2</span> Extracting information from combined variables using <code>tidyr::extract</code></h3>
<p>The column <code>expt.and.time</code> violates one of our principles of tidy data: “Each variable must have its own column.”. This column conflates two different types of information – the experiment type and the time point of the measurement. Our next task is to split this information up into two new variables, which will help to facilitate downstream plotting and analysis.</p>
<p>One complicating factor is that the different experiments/time combinations have different naming conventions:</p>
<ul>
<li><p>The “alpha” and “elu” experiments are of the form “alpha0”, “alpha7”, “elu0”, “elu30”, etc. In this case, the first part of the string gives the experiment type (either alpha or elu) and the following digits give the time point.</p></li>
<li><p>In the “cdc15” and “cdc28” experiments the convention is slightly different; they are of the form “cdc15_0”, “cdc15_10”, “cdc28_0”, “cdc28_10”, etc. Here the part of the string before the underscore gives the experiment type, and the digits after the underscore give the time point.</p></li>
</ul>
<p>Because of the differences in naming conventions, we will find it easiest to break up <code>spellman.long</code> into a series of sub-data sets corresponding to each experiment type in order to extract out the experiment and time information. After processing each data subset separately, we will join the modified sub-data frames back together.</p>
</div>
<div id="subsetting-rows" class="section level3">
<h3><span class="header-section-number">8.6.3</span> Subsetting rows</h3>
<p>Let’s start by getting just the rows corresponding to the “alpha” experiment/times. Here we use <code>dplyr::filter</code> in combination with <code>stringr::str_detect</code> to get all those rows in which the <code>expt.and.time</code> variable contains the string “alpha”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha.long &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;alpha&quot;</span>))

<span class="co"># look at the new data frame</span>
<span class="kw">dim</span>(alpha.long)
<span class="co">#&gt; [1] 111204      5</span>
<span class="kw">head</span>(alpha.long, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">#&gt; # A tibble: 10 x 5</span>
<span class="co">#&gt;    gene   std.name description                    expt.and.time expression</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                          &lt;chr&gt;              &lt;dbl&gt;</span>
<span class="co">#&gt;  1 YAL00… TFC3     Subunit of RNA polymerase III… alpha0             -0.15</span>
<span class="co">#&gt;  2 YAL00… VPS8     Membrane-binding component of… alpha0             -0.11</span>
<span class="co">#&gt;  3 YAL00… EFB1     Translation elongation factor… alpha0             -0.14</span>
<span class="co">#&gt;  4 YAL00… &lt;NA&gt;     Dubious open reading frame; u… alpha0             -0.02</span>
<span class="co">#&gt;  5 YAL00… SSA1     ATPase involved in protein fo… alpha0             -0.05</span>
<span class="co">#&gt;  6 YAL00… ERP2     Member of the p24 family invo… alpha0             -0.6 </span>
<span class="co">#&gt;  7 YAL00… FUN14    Integral mitochondrial outer … alpha0             -0.28</span>
<span class="co">#&gt;  8 YAL00… SPO7     Putative regulatory subunit o… alpha0             -0.03</span>
<span class="co">#&gt;  9 YAL01… MDM10    Subunit of both the ERMES and… alpha0             -0.05</span>
<span class="co">#&gt; 10 YAL01… SWC3     Protein of unknown function; … alpha0             -0.31</span></code></pre></div>
</div>
<div id="splitting-columns" class="section level3">
<h3><span class="header-section-number">8.6.4</span> Splitting columns</h3>
<p>Having subsetted the data, we can now split <code>expt.and.time</code> into two new variables – <code>expt</code> and <code>time</code>. To do this we use <code>tidyr::extract</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha.long <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time,     <span class="co"># column we&#39;re extracting from</span>
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="co"># new columns we&#39;re creating</span>
                 <span class="dt">regex=</span><span class="st">&quot;(alpha)([[:digit:]]+)&quot;</span>,  <span class="co"># regexp (see below)</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)      <span class="co"># automatically convert column types</span>

<span class="co"># </span><span class="al">NOTE</span><span class="co">: I&#39;m being explict about saying tidyr::extract because the</span>
<span class="co"># magrittr package defines a different extract function</span></code></pre></div>
<p>Let’s take a moment to look at the <code>regex</code> argument to <code>extract</code> – <code>regex=&quot;(alpha)([[:digit:]]+)&quot;</code>. The regex is specified as a character string. Each part we want to match <em>and</em> extract is surround by parentheses. In this case we have two sets of parentheses corresponding to the two matches we want to make. The first part of the regex is <code>(alpha)</code>; here we’re looking to make an exact match to the string “alpha”. The second part of the regex reads <code>([[:digit:]]+)</code>. <code>[[:digit:]]</code> indicates we’re looking for a numeric digit. The <code>+</code> after <code>[[:digit:]]</code> indicates that we want to match <em>one or more</em> digits (i.e. to get a match we need to find at least one digit, but more than one digit should also be a match).</p>
<p>Let’s take a look at the new version of <code>alpha.long</code> following application of <code>extract</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(alpha.long, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">#&gt; # A tibble: 10 x 6</span>
<span class="co">#&gt;    gene   std.name description                      expt   time expression</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                            &lt;chr&gt; &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 YAL00… TFC3     Subunit of RNA polymerase III t… alpha     0      -0.15</span>
<span class="co">#&gt;  2 YAL00… VPS8     Membrane-binding component of t… alpha     0      -0.11</span>
<span class="co">#&gt;  3 YAL00… EFB1     Translation elongation factor 1… alpha     0      -0.14</span>
<span class="co">#&gt;  4 YAL00… &lt;NA&gt;     Dubious open reading frame; unl… alpha     0      -0.02</span>
<span class="co">#&gt;  5 YAL00… SSA1     ATPase involved in protein fold… alpha     0      -0.05</span>
<span class="co">#&gt;  6 YAL00… ERP2     Member of the p24 family involv… alpha     0      -0.6 </span>
<span class="co">#&gt;  7 YAL00… FUN14    Integral mitochondrial outer me… alpha     0      -0.28</span>
<span class="co">#&gt;  8 YAL00… SPO7     Putative regulatory subunit of … alpha     0      -0.03</span>
<span class="co">#&gt;  9 YAL01… MDM10    Subunit of both the ERMES and t… alpha     0      -0.05</span>
<span class="co">#&gt; 10 YAL01… SWC3     Protein of unknown function; co… alpha     0      -0.31</span></code></pre></div>
<p>Notice our two new variables, both of which have appropriate types!</p>
<p>A data frame for the elutriation data can be created similarly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">elu.long &lt;-
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;elu&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time,     <span class="co"># column we&#39;re extracting from</span>
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="co"># new columns we&#39;re creating</span>
                 <span class="dt">regex=</span><span class="st">&quot;(elu)([[:digit:]]+)&quot;</span>,  <span class="co"># regexp (see below)</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)      <span class="co"># automatically convert column types</span></code></pre></div>
<div id="a-fancier-regex-for-the-cdc-experiments" class="section level4">
<h4><span class="header-section-number">8.6.4.1</span> A fancier regex for the cdc experiments</h4>
<p>Now let’s process the cdc experiments (cdc15 and cdc28). As before we extract the corresponding rows of the data frame using <code>filter</code> and <code>str_detect</code>. We then split <code>expt.and.time</code> using <code>tidyr::extract</code>. In this case we carry out the two steps in a single code block using pipes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cdc.long &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># both cdc15 and cdc28 contain &quot;cdc&quot; as a sub-string</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;cdc&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time, 
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), 
                 <span class="dt">regex=</span><span class="st">&quot;(cdc15|cdc28)_([[:digit:]]+)&quot;</span>, <span class="co"># note the fancier regex</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>The regex – <code>&quot;(cdc15|cdc28)_([[:digit:]]+)&quot;</code> – is slightly fancier in this example. As before there are two parts we’re extracting: <code>(cdc15|cdc28)</code> and <code>([[:digit:]]+)</code>. The first parenthesized regexp is an “OR” – i.e. match “cdc15” <em>or</em> “cdc28”. The second parenthesized regexp is the same as we saw previously. Separating the two parenthesized regexps is an underscore (<code>_</code>). The underscore isn’t parenthesized because we only want to use it to make a match <em>not</em> to extract the corresponding match.</p>
</div>
</div>
<div id="combining-data-frames" class="section level3">
<h3><span class="header-section-number">8.6.5</span> Combining data frames</h3>
<p>If you have two or more data frames with identical columns, the rows of the data frames can be combined into a single data frame using <code>rbind</code> (defined in the <code>base</code> package). For example, to reassemble the <code>alpha.long</code>, <code>elu.long</code>, and <code>cdc.long</code> data frames into a single data frame we do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final &lt;-<span class="st"> </span><span class="kw">rbind</span>(alpha.long, elu.long, cdc.long)

<span class="co"># check the dimensions of the new data frame</span>
<span class="kw">dim</span>(spellman.final)
<span class="co">#&gt; [1] 450994      6</span></code></pre></div>
</div>
<div id="sorting-data-frame-rows" class="section level3">
<h3><span class="header-section-number">8.6.6</span> Sorting data frame rows</h3>
<p>Currently the <code>spellman.final</code> data frame is sorted by time point and experiment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(spellman.final, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">#&gt; # A tibble: 10 x 6</span>
<span class="co">#&gt;    gene   std.name description                      expt   time expression</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                            &lt;chr&gt; &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 YAL00… TFC3     Subunit of RNA polymerase III t… alpha     0      -0.15</span>
<span class="co">#&gt;  2 YAL00… VPS8     Membrane-binding component of t… alpha     0      -0.11</span>
<span class="co">#&gt;  3 YAL00… EFB1     Translation elongation factor 1… alpha     0      -0.14</span>
<span class="co">#&gt;  4 YAL00… &lt;NA&gt;     Dubious open reading frame; unl… alpha     0      -0.02</span>
<span class="co">#&gt;  5 YAL00… SSA1     ATPase involved in protein fold… alpha     0      -0.05</span>
<span class="co">#&gt;  6 YAL00… ERP2     Member of the p24 family involv… alpha     0      -0.6 </span>
<span class="co">#&gt;  7 YAL00… FUN14    Integral mitochondrial outer me… alpha     0      -0.28</span>
<span class="co">#&gt;  8 YAL00… SPO7     Putative regulatory subunit of … alpha     0      -0.03</span>
<span class="co">#&gt;  9 YAL01… MDM10    Subunit of both the ERMES and t… alpha     0      -0.05</span>
<span class="co">#&gt; 10 YAL01… SWC3     Protein of unknown function; co… alpha     0      -0.31</span></code></pre></div>
<p>It might be useful instead to sort by gene and experiment. To do this we can use <code>dplyr::arrange</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(gene, expt)

<span class="co"># look again at the rearranged data</span>
<span class="kw">head</span>(spellman.final, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">#&gt; # A tibble: 10 x 6</span>
<span class="co">#&gt;    gene   std.name description                      expt   time expression</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                            &lt;chr&gt; &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 YAL00… TFC3     Subunit of RNA polymerase III t… alpha     0      -0.15</span>
<span class="co">#&gt;  2 YAL00… TFC3     Subunit of RNA polymerase III t… alpha     7      -0.15</span>
<span class="co">#&gt;  3 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    14      -0.21</span>
<span class="co">#&gt;  4 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    21       0.17</span>
<span class="co">#&gt;  5 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    28      -0.42</span>
<span class="co">#&gt;  6 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    35      -0.44</span>
<span class="co">#&gt;  7 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    42      -0.15</span>
<span class="co">#&gt;  8 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    49       0.24</span>
<span class="co">#&gt;  9 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    56      -0.1 </span>
<span class="co">#&gt; 10 YAL00… TFC3     Subunit of RNA polymerase III t… alpha    63      NA</span></code></pre></div>
</div>
</div>
<div id="using-your-tidy-data" class="section level2">
<h2><span class="header-section-number">8.7</span> Using your tidy data</h2>
<p>Whew – that was a fair amount of work to tidy our data! But having done so we can now carry out a wide variety of very powerful analyses.</p>
<div id="visualizing-gene-expression-time-series" class="section level3">
<h3><span class="header-section-number">8.7.1</span> Visualizing gene expression time series</h3>
<p>Let’s start by walking through a series of visualizations of gene expression time series. Each plot will show the expression of one or more genes, at different time points, in one or more experimental conditions. Our initial visualizations exploit the “long” versions of the tidy data.</p>
<p>First a single gene in a single experimental condition:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>, gene <span class="op">==</span><span class="st"> &quot;YAL022C&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression of YAL022C&quot;</span>)</code></pre></div>
<p><img src="bio304-book_files/figure-html/unnamed-chunk-272-1.png" width="576" /></p>
<p>We can easily modify the above code block to visualize the expression of multiple genes of interest:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes.of.interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;YAL022C&quot;</span>, <span class="st">&quot;YAR018C&quot;</span>, <span class="st">&quot;YGR188C&quot;</span>)

spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest, expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of multiple genes</span><span class="ch">\n</span><span class="st">following synchronization by alpha factor&quot;</span>)</code></pre></div>
<p><img src="bio304-book_files/figure-html/unnamed-chunk-273-1.png" width="576" /></p>
<p>By employing <code>facet_wrap()</code> we can visualize the relationship between this set of genes in each of the experiment types:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of Multiple Genes</span><span class="ch">\n</span><span class="st">Across experiments&quot;</span>) </code></pre></div>
<p><img src="bio304-book_files/figure-html/unnamed-chunk-274-1.png" width="576" /></p>
<p>The different experimental treatments were carried out for varying lengths of time due to the differences in their physiological effects. Plotting them all on the same time scale can obscure that patterns of oscillation we might be interested in, so let’s modify our code block so that plots that share the same y-axis, but have differently scaled x-axes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of Multiple Genes</span><span class="ch">\n</span><span class="st">Across experiments&quot;</span>) </code></pre></div>
<p><img src="bio304-book_files/figure-html/unnamed-chunk-275-1.png" width="576" /></p>
</div>
<div id="finding-the-most-variable-genes" class="section level3">
<h3><span class="header-section-number">8.7.2</span> Finding the most variable genes</h3>
<p>When dealing with vary large data sets, one ad hoc filtering criteria that is often employed is to focus on those variables that exhibit that greatest variation (variation is measure of the spread of data; we will give a precise definition in a later lecture). To do this, we first need to order our variables (genes) by their variation. Let’s see how we can accomplish this using our long data frame:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">by.variance &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(expression.var))

<span class="kw">head</span>(by.variance)
<span class="co">#&gt; # A tibble: 6 x 2</span>
<span class="co">#&gt;   gene    expression.var</span>
<span class="co">#&gt;   &lt;chr&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt; 1 YLR286C          2.157</span>
<span class="co">#&gt; 2 YNR067C          1.733</span>
<span class="co">#&gt; 3 YNL327W          1.653</span>
<span class="co">#&gt; 4 YGL028C          1.571</span>
<span class="co">#&gt; 5 YHL028W          1.521</span>
<span class="co">#&gt; 6 YKL164C          1.515</span></code></pre></div>
<p>The code above calculates the variance of each gene but ignores the fact that we have different experimental conditions. To take into account the experimental design of the data at hand, let’s calculate the average variance across the experimental conditions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">by.avg.variance &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene, expt) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">avg.expression.var =</span> <span class="kw">mean</span>(expression.var)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(avg.expression.var))

<span class="kw">head</span>(by.avg.variance)
<span class="co">#&gt; # A tibble: 6 x 2</span>
<span class="co">#&gt;   gene    avg.expression.var</span>
<span class="co">#&gt;   &lt;chr&gt;                &lt;dbl&gt;</span>
<span class="co">#&gt; 1 YFR014C              3.579</span>
<span class="co">#&gt; 2 YFR053C              2.377</span>
<span class="co">#&gt; 3 YBL032W              2.299</span>
<span class="co">#&gt; 4 YDR274C              2.173</span>
<span class="co">#&gt; 5 YLR286C              2.128</span>
<span class="co">#&gt; 6 YMR206W              1.937</span></code></pre></div>
<p>Based on the average experession variance across experimental conditions, let’s get the names of the 1000 most variable genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top.genes.1k &lt;-<span class="st"> </span>by.avg.variance[<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>,]<span class="op">$</span>gene

<span class="kw">head</span>(top.genes.1k)
<span class="co">#&gt; [1] &quot;YFR014C&quot; &quot;YFR053C&quot; &quot;YBL032W&quot; &quot;YDR274C&quot; &quot;YLR286C&quot; &quot;YMR206W&quot;</span></code></pre></div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-the-dplyr-package.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
